activerecord/lib/active_record/associations/preloader/association.rb

```diff
@@ -115,7 +115,16 @@ def load_records(&block)
             # Make several smaller queries if necessary or make one query if the adapter supports it
             slices = owner_keys.each_slice(klass.connection.in_clause_length || owner_keys.size)
             @preloaded_records = slices.flat_map do |slice|
-              records_for(slice).load(&block)
+              records = records_for(slice)
```

Comment:
I saw the test case in https://github.com/rails/rails/pull/28548/commits/6457950237447dbb4e8e0e445df99fee7980ec1c.
What `records_for` returns an array means that the user override the `records_for`.
If so, passing `&block` to `records_for` ease to handle it themself.